enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,690 @@
+which is based on RHEL For Edge which features ostree and as such
+provides rollbacks to go back to previous ostree deployment.
+Even though, OpenShift does not support downgrade or rollback,
+MicroShift must support it in some form. Explicit downgrade 
+will not be supported, and rollback will be supported only when
+newer ostree commits is unhealthy and backup consistent with previously
+ran MicroShift is present.
+
+To allow for such operations, we need to define how we'll
+achieve that goal. We can define several areas we need to focus on:
+backing up and restoring MicroShift's data, handling version changes
+and its consequences such as migrating underlying data between schema
+versions, defining a mechanism for allowing or blocking upgrades between
+certain version of MicroShift.
+
+### User Stories
+
+* As a MicroShift administrator, I want to safely update MicroShift
+  so that I can get bug fixes, new features, and security patches.
+* As a MicroShift administrator, I want automated backups of 
+  MicroShift data so in case of a greenboot failure system will rollback
+  and backup will be restored.
+
+### Goals
+
+Goal of the enhancement is to describe implementation roadmap for
+integrating MicroShift with ostree and greenboot in order to provide
+functionality to:
+- Safely update MicroShift version (by backing up the data and
+  restoring it in case of rollback)
+- Migrating internal data (like Kubernetes storage or etcd schema) to
+  newer version
+- Block upgrades in case of version bump being too big
+
+Design aims to implement following principles:
+- Keep it simple, optimize later
+- MicroShift does not own the OS or host
+- MicroShift and all its components are versioned, upgraded, and rolled back together
+- Be defensive, fail fast
+- Rely on outside intervention as a last resort
+
+### Non-Goals
+
+* Building allowed/blocked version migration graph
+* Handling readiness, and backup and rollback of 3rd party applications
+  (although end user documentation should be provided)
+* Defining updateability for non-ostree systems is left to a future enhancement
+* Protecting against data corruption - we rely on file system 
+  to maintain the backed up file integrity
+
+## Proposal
+
+### Workflow Description
+
+#### User perspective
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots to new commit
+1. Operating System, greenboot, and MicroShift take actions without any additional intervention
+
+#### Implementation perspective
+
+**First boot into commit with MicroShift**
+
+1. First ostree commit is installed
+1. First ostree commit boots
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
```

Comment:
The backup will be generated on the next boot, but if it fails then there is no backup to restore. This would be a new ostree commit falling into the *Failed upgrade* use case.

Reply:
Not sure what you mean. We also have "Failed first boot into commit with MicroShift" below

Reply:
I know what you meant now, I'll think about this